### PR TITLE
linagora/lgs/linshare/products/linshare-mobile-apps#534 LA-534 unit test for domain layer

### PR DESCRIPTION
**Pull request by manhduc-nguyen**
Implement unit test for:
- CreatePermanentTokenInteractor
- GetCredentialInteractor
- UploadFileInteractor

### CODE REVIEW

**Comment by manhduc-nguyen** [comment]
@hoangdat pls review this PR.

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    CreatePermanentTokenInteractor createPermanentTokenInteractor;
+    MockAuthenticationRepository authenticationRepository;
+    MockTokenRepository tokenRepository;
+    MockCredentialRepository credentialRepository;
+
+    setUp(() {
+      authenticationRepository = MockAuthenticationRepository();
+      tokenRepository = MockTokenRepository();
+      credentialRepository = MockCredentialRepository();
+      createPermanentTokenInteractor = CreatePermanentTokenInteractor(authenticationRepository, tokenRepository, credentialRepository);
```
Comment (hoangdat) [inline domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:55]: too long line

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    test('createPermanentTokenInteractorShouldFailureWithInternalServerError', () async {
+      when(authenticationRepository.createPermanentToken(linShareBaseUrl, userName1, password1))
+          .thenThrow(InternalServerError());
+      final result = await createPermanentTokenInteractor.execute(linShareBaseUrl, userName1, password1);
+      expect(result, Left<Failure, Success>(AuthenticationFailure(InternalServerError())));
+    });
+
+    test('createPermanentTokenInteractorShouldFailureWithMissingRequiredFieldError', () async {
+      when(authenticationRepository.createPermanentToken(linShareBaseUrl, UserName(''), password1))
+          .thenThrow(MissingRequiredFields());
```
Comment (hoangdat) [inline domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:109]: Really, real function will throw this Error?

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    test('createPermanentTokenInteractorShouldFailureWithUnknownError', () async {
+      when(authenticationRepository.createPermanentToken(linShareBaseUrl, userName1, password1))
+          .thenThrow(UnknownError('unknown error'));
+      final result = await createPermanentTokenInteractor.execute(linShareBaseUrl, userName1, password1);
+      expect(result, Left<Failure, Success>(AuthenticationFailure(UnknownError('unknown error'))));
+    });
+
+    test('createPermanentTokenInteractorShouldFailureWithInternalServerError', () async {
+      when(authenticationRepository.createPermanentToken(linShareBaseUrl, userName1, password1))
+          .thenThrow(InternalServerError());
```
Comment (hoangdat) [inline domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:102]: Really?, as I understand, with this kind of error, we will throw `UnknownError`? Am I right?

File: domain/lib/src/usecases/authentication/get_credential_interactor.dart [hunk 4]
```diff
@@ -43,7 +43,11 @@ class GetCredentialInteractor {
     try {
       final token = await tokenRepository.getToken();
       final baseUrl = await credentialRepository.getBaseUrl();
-      return Right(CredentialViewState(token, baseUrl));
+      if (baseUrl.origin.isNotEmpty) {
+        return Right(CredentialViewState(token, baseUrl));
+      } else {
+        return Left(CredentialFailure(BadCredentials()));
+      }
```
Comment (hoangdat) [inline domain/lib/src/usecases/authentication/get_credential_interactor.dart:50]: how about if `token` is empty? please, use the other way, not `if-else`

File: domain/test/usecases/authentication/get_credential_interactor_test.dart [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    MockTokenRepository tokenRepository;
+    MockCredentialRepository credentialRepository;
+
+    setUp(() {
+      tokenRepository = MockTokenRepository();
+      credentialRepository = MockCredentialRepository();
+      getCredentialInteractor = GetCredentialInteractor(tokenRepository, credentialRepository);
+    });
+
+    test('getCredentialInteractorShouldReturnSuccessWithCorrectData', () async {
```
Comment (hoangdat) [inline domain/test/usecases/authentication/get_credential_interactor_test.dart:55]: one remark for the future: we can make the test case as a human language, it is readable, why we don't use it?

File: domain/lib/src/usecases/myspace/upload_file_interactor.dart [hunk 6]
```diff
@@ -55,8 +55,7 @@ class UploadFileInteractor {
           return Left(UploadFileFailure(Exception()));
         }, (success) {
           if (success is FileUploadProgress) {
```
Comment (hoangdat) [inline domain/lib/src/usecases/myspace/upload_file_interactor.dart:57]: TBH, I am not happy with it `if-else`, I like the way we need to fight with the old way. Fighting!!!

File: domain/test/usecases/authentication/upload_file_interactor_test.dart [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+          emitsInOrder([
+            Right<Failure, Success>(PreparingUpload(fileInfo1)),
+            Left<Failure, Success>(UploadFileFailure(exception))
+          ]));
+    });
+
+    test('uploadFileInteractorShouldFailureWithWrongToken', () async {
+      final wrongToken = Token('token', TokenId('uuid'));
+      when(tokenRepository.getToken()).thenAnswer((_) async => wrongToken);
+      when(credentialRepository.getBaseUrl()).thenAnswer((_) async => linShareBaseUrl);
```
Comment (hoangdat) [inline domain/test/usecases/authentication/upload_file_interactor_test.dart:104]: How about if `ServerNotFound`, `TimeoutConnect`, ...or any error with server?

**Comment by hoangdat** [comment]
- We have the same error with CI. 

- Are your sure your test case in `domain` is run? What happen with CI if some of these is failed?

**Comment by tnstung** [comment]
> * We have the same error with CI.
> * Are your sure your test case in `domain` is run? What happen with CI if some of these is failed?

Do you both succeed to run those test cases in your local?

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart (hunk 2 above)
Comment (manhduc-nguyen) [reply domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:109]: Screen Shot 2020-11-06 at 09 54 15

Yes, we have that in our api document. I also tested with postman, same response is returned when I removed `label` data in `createPermanentToken` api

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart (hunk 3 above)
Comment (manhduc-nguyen) [reply domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:102]: Oh I think this test is correct but I missed logic code for this case in `AuthenticationDataSource`. Let me add that.

File: domain/test/usecases/authentication/create_permanent_token_interactor_test.dart (hunk 2 above)
Comment (manhduc-nguyen) [reply domain/test/usecases/authentication/create_permanent_token_interactor_test.dart:109]: same, let me add this case in `AuthenticationDataSource`.

File: domain/test/usecases/authentication/upload_file_interactor_test.dart (hunk 7 above)
Comment (manhduc-nguyen) [reply domain/test/usecases/authentication/upload_file_interactor_test.dart:104]: Ok I will add this later when I update `datasource`

**Pull request closed without merging**